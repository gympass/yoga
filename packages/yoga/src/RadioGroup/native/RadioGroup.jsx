import React, { useState } from 'react';
import { View } from 'react-native';
import { string, func, oneOfType, number, bool, node } from 'prop-types';
import styled from 'styled-components';

import RadioGroupContext from '../RadioGroupContext';
import RadioButton from './Button/RadioButton';

const Group = styled.View(
  () => `
    align-self: center;
    position: relative;
    flex-direction: row;
  `,
);

const GrayLine = styled.View(
  ({
    theme: {
      yoga: {
        components: { radiogroup },
      },
    },
  }) => `
    position: absolute;
    height: 100%;
    border-radius: ${radiogroup.button.border.radius}px;
    border-width: ${radiogroup.button.border.width}px;
    border-color: ${radiogroup.button.border.color};
  `,
);

const RadioGroup = ({
  onChange = () => {},
  selectedValue = '',
  small = false,
  children,
  ...rest
}) => {
  const [groupSize, setGroupSize] = useState(0);

  const isButton = React.Children.map(children, child => child.type).every(
    child => child === RadioButton,
  );

  return (
    <RadioGroupContext.Provider
      value={{
        selectedValue,
        onChange,
        small,
        ...rest,
      }}
    >
      <View>
        <Group
          {...rest}
          onLayout={({
            nativeEvent: {
              layout: { width },
            },
          }) => setGroupSize(width)}
        >
          {Boolean(isButton && groupSize) && (
            <GrayLine style={{ width: groupSize }} />
          )}

          {children}
        </Group>
      </View>
    </RadioGroupContext.Provider>
  );
};

RadioGroup.propTypes = {
  onChange: func,
  selectedValue: oneOfType([string, number]),
  small: bool,
  children: node.isRequired,
};

export default RadioGroup;
