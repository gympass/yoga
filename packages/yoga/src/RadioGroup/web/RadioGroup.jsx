import React from 'react';
import { string, func, oneOfType, number, bool, node, shape } from 'prop-types';
import styled from 'styled-components';

import RadioGroupContext from '../RadioGroupContext';

const Group = styled.div.attrs({
  role: 'radiogroup',
})`
  display: flex;

  width: min-content;

  ${({ full }) => `
    ${full ? `width: 100%;` : ``}
  `}
`;

const RadioGroup = ({
  name = '',
  onChange = () => {},
  selectedValue = '',
  small = false,
  full = false,
  children = null,
  style = {},
  ...rest
}) => (
  <RadioGroupContext.Provider
    value={{
      selectedValue,
      name,
      onChange,
      small,
      ...rest,
    }}
  >
    <Group full={full} style={style} {...rest}>
      {children}
    </Group>
  </RadioGroupContext.Provider>
);

RadioGroup.propTypes = {
  name: string,
  onChange: func,
  selectedValue: oneOfType([string, number]),
  small: bool,
  full: bool,
  children: node,
  style: shape({}),
};

export default RadioGroup;
