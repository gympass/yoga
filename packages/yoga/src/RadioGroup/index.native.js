import { RadioGroup, RadioButton, Radio } from './native';

const ExportRadioGroup = Object.assign(RadioGroup, {
  Button: RadioButton,
  Radio,
});

export default ExportRadioGroup;
