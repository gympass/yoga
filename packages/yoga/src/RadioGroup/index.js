import { RadioGroup, RadioButton, Radio } from './web';

const ExportRadioGroup = Object.assign(RadioGroup, {
  Button: RadioButton,
  Radio,
});

export default ExportRadioGroup;
