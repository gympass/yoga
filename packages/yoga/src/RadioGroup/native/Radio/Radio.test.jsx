import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';

import RadioGroup from '../..';
import ThemeProvider from '../../../Theme';
import RadioGroupContext from '../../RadioGroupContext';

describe('<RadioGroup />', () => {
  describe('<RadioGroup.Radio />', () => {
    describe('Snapshots', () => {
      it('should match snapshot with default radio', () => {
        const { container, toJSON } = render(
          <ThemeProvider>
            <RadioGroup.Radio />
          </ThemeProvider>,
        );

        expect(toJSON(container)).toMatchSnapshot();
      });

      it('should match snapshot with disabled radio', () => {
        const { container, toJSON } = render(
          <ThemeProvider>
            <RadioGroup.Radio disabled />
          </ThemeProvider>,
        );

        expect(toJSON(container)).toMatchSnapshot();
      });

      it('should match snapshot with default radio', () => {
        const { container, toJSON } = render(
          <ThemeProvider>
            <RadioGroup.Radio />
          </ThemeProvider>,
        );

        expect(toJSON(container)).toMatchSnapshot();
      });

      it('should match snapshot with default radio pressed', () => {
        const onChangeMock = jest.fn();
        const { container, getByTestId, toJSON } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Radio testID="radio" />
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByTestId(/^radio/);

        fireEvent.press(radio);

        expect(toJSON(container)).toMatchSnapshot();
      });
    });

    describe('Props', () => {
      it('checked should be the selectedValue provided from Context', () => {
        const { getAllByTestId } = render(
          <RadioGroupContext.Provider value={{ selectedValue: 'Radio 2' }}>
            <ThemeProvider>
              <RadioGroup.Radio testID="radio" value="Radio 1" />
              <RadioGroup.Radio testID="radio" value="Radio 2" />
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const [radio, radio2] = getAllByTestId(/^radio/);

        expect(radio.props.checked).toBe(false);
        expect(radio2.props.checked).toBe(true);
      });
    });

    describe('Events', () => {
      it('should call onchange from Context', () => {
        const onChangeMock = jest.fn();

        const { getByTestId } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Radio testID="radio" />
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByTestId(/^radio/);

        fireEvent.press(radio);

        expect(onChangeMock).toHaveBeenCalled();
      });

      it('value should be as prop value', () => {
        const onChangeMock = jest.fn();
        const { getByTestId } = render(
          <RadioGroupContext.Provider value={{ onChange: onChangeMock }}>
            <ThemeProvider>
              <RadioGroup.Radio value="radio1" testID="radio" />
            </ThemeProvider>
          </RadioGroupContext.Provider>,
        );

        const radio = getByTestId(/^radio/);

        fireEvent.press(radio);

        expect(onChangeMock).toHaveBeenCalledWith({
          value: 'radio1',
        });
      });
    });
  });
});
