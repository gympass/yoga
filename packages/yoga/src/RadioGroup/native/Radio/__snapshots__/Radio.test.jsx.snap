// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> <RadioGroup.Radio /> Snapshots should match snapshot with default radio 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  checked={false}
  disabled={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pressed={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9898A6",
        "borderRadius": 9999,
        "borderStyle": "solid",
        "borderWidth": 2,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
/>
`;

exports[`<RadioGroup /> <RadioGroup.Radio /> Snapshots should match snapshot with default radio 2`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  checked={false}
  disabled={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pressed={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9898A6",
        "borderRadius": 9999,
        "borderStyle": "solid",
        "borderWidth": 2,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
/>
`;

exports[`<RadioGroup /> <RadioGroup.Radio /> Snapshots should match snapshot with default radio pressed 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  checked={false}
  disabled={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pressed={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#9898A6",
        "borderRadius": 9999,
        "borderStyle": "solid",
        "borderWidth": 2,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
  testID="radio"
/>
`;

exports[`<RadioGroup /> <RadioGroup.Radio /> Snapshots should match snapshot with disabled radio 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": true,
    }
  }
  accessible={true}
  checked={false}
  disabled={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  pressed={false}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "borderColor": "#F5F5FA",
        "borderRadius": 9999,
        "borderStyle": "solid",
        "borderWidth": 2,
        "height": 20,
        "justifyContent": "center",
        "width": 20,
      },
    ]
  }
/>
`;
